File: src/util/entities/ConnectedAccount.ts
/*
	Spacebar: A FOSS re-implementation and extension of the Discord.com backend.
	Copyright (C) 2023 Spacebar and Spacebar Contributors

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU Affero General Public License as published
	by the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU Affero General Public License for more details.

	You should have received a copy of the GNU Affero General Public License
	along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import { Column, Entity, JoinColumn, ManyToOne, RelationId } from "typeorm";
import { ConnectedAccountTokenData } from "../interfaces";
import { BaseClass } from "./BaseClass";
import { User } from "./User";

@Entity({
	name: "connected_accounts",
})
export class ConnectedAccount extends BaseClass {
	@Column()
	external_id: string;

	@Column({ nullable: true })
	@RelationId((account: ConnectedAccount) => account.user)
	user_id: string;

	@JoinColumn({ name: "user_id" })
	@ManyToOne(() => User, {
		onDelete: "CASCADE",
	})
	user: User;

	@Column({ select: false })
	friend_sync?: boolean = false;

	@Column()
	name: string;

	@Column({ select: false })
	revoked?: boolean = false;

	@Column({ select: false })
	show_activity?: number = 0;

	@Column()
	type: string;

	@Column()
	verified?: boolean = true;

	@Column({ select: false })
	visibility?: number = 0;

	@Column({ type: "simple-array" })
	integrations?: string[] = [];

	@Column({ type: "simple-json", name: "metadata", nullable: true })
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	metadata_?: any;

	@Column()
	metadata_visibility?: number = 0;

	@Column()
	two_way_link?: boolean = false;

	@Column({ select: false, nullable: true, type: "simple-json" })
	token_data?: ConnectedAccountTokenData | null;

	async revoke() {
		this.revoked = true;
		this.token_data = null;
		await this.save();
	}
}

File: extra/admin-api/Models/Spacebar.Models.Db/Models/ConnectedAccount.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace Spacebar.Models.Db.Models;

[Table("connected_accounts")]
public partial class ConnectedAccount
{
    [Key]
    [Column("id", TypeName = "character varying")]
    public string Id { get; set; } = null!;

    [Column("external_id", TypeName = "character varying")]
    public string ExternalId { get; set; } = null!;

    [Column("user_id", TypeName = "character varying")]
    public string? UserId { get; set; }

    [Column("friend_sync")]
    public bool FriendSync { get; set; }

    [Column("name", TypeName = "character varying")]
    public string Name { get; set; } = null!;

    [Column("revoked")]
    public bool Revoked { get; set; }

    [Column("show_activity")]
    public int ShowActivity { get; set; }

    [Column("type", TypeName = "character varying")]
    public string Type { get; set; } = null!;

    [Column("verified")]
    public bool Verified { get; set; }

    [Column("visibility")]
    public int Visibility { get; set; }

    [Column("integrations")]
    public string Integrations { get; set; } = null!;

    [Column("metadata")]
    public string? Metadata { get; set; }

    [Column("metadata_visibility")]
    public int MetadataVisibility { get; set; }

    [Column("two_way_link")]
    public bool TwoWayLink { get; set; }

    [Column("token_data")]
    public string? TokenData { get; set; }

    [ForeignKey("UserId")]
    [InverseProperty("ConnectedAccounts")]
    public virtual User? User { get; set; }
}
